pkg/buildkit/archive/archive.go

```diff
@@ -194,25 +201,49 @@ func extract(fp string, ct mimeType, dst string) error {
 			continue
 		}
 
-		target := filepath.Join(dst, header.Name)
+		target, err := sanitizeExtractPath(dst, header.Name)
+		if err != nil {
+			return err
+		}
 
 		switch header.Typeflag {
 		case tar.TypeDir:
 			if err := os.MkdirAll(target, 0755); err != nil {
 				return err
 			}
 		case tar.TypeReg:
-			f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
-			if err != nil {
+			if err = copyRegularFile(target, tr, header.Mode); err != nil {
 				return err
 			}
+		}
+	}
+}
 
-			_, err = io.Copy(f, tr)
-			f.Close()
+func sanitizeExtractPath(destination, filename string) (string, error) {
+	destPath := filepath.Join(destination, filename)
+	if !strings.HasPrefix(destPath, filepath.Clean(destination)) {
+		return "", fmt.Errorf("content filepath tainted: %s", destPath)
+	}
 
-			if err != nil {
-				return err
+	return destPath, nil
+}
+
+func copyRegularFile(target string, tr *tar.Reader, mode int64) error {
+	f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(mode))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for {
+		if _, err = io.CopyN(f, tr, 1024); err != nil {
```

Comment:
Same, just curious why is this better than `io.Copy`?

Reply:
https://www.cvedetails.com/cwe-details/409/Improper-Handling-of-Highly-Compressed-Data-Data-Amplificat.html

This issue was raised by the gosec linter as G110